jenkins/ci.suse.de/pipelines/openstack-ardana-gerrit.Jenkinsfile

```diff
@@ -23,74 +23,136 @@ pipeline {
     stage('Setup workspace') {
       steps {
         script {
-          currentBuild.displayName = "#${BUILD_NUMBER}: ${gerrit_change_ids}"
-
           sh('''
             git clone $git_automation_repo --branch $git_automation_branch automation-git
           ''')
 
           ardana_lib = load "$WORKSPACE/automation-git/jenkins/ci.suse.de/pipelines/openstack-ardana.groovy"
           ardana_lib.load_extra_params_as_vars(extra_params)
+
+          if (GERRIT_CHANGE_NUMBER == '') {
+            error("Empty 'GERRIT_CHANGE_NUMBER' parameter value.")
+          }
+
+          if (env.GERRIT_PATCHSET_NUMBER == null) {
+            // Extract the current patchset number from the Gerrit change, if missing
+            env.GERRIT_PATCHSET_NUMBER = sh (
+              returnStdout: true,
+              script: '''
+                automation-git/scripts/jenkins/ardana/gerrit/gerrit_get.py \
+                  --attr patchset \
+                  ${GERRIT_CHANGE_NUMBER}
+              '''
+            ).trim()
+          }
+
+          currentBuild.displayName = "#${BUILD_NUMBER}: $GERRIT_CHANGE_NUMBER/$GERRIT_PATCHSET_NUMBER ($gerrit_context)"
         }
+      }
+    }
 
-        sh('''
-          if [ -n "$GERRIT_CHANGE_NUMBER" ] ; then
-            # Post reviews only for jobs triggered by Gerrit
-            automation-git/scripts/jenkins/ardana/gerrit/gerrit_review.py \
-              --vote 0 \
-              --label 'Verified' \
-              --message "
-Started build (${JOB_NAME}): ${BUILD_URL}
+    stage('supersede running jobs') {
+      steps {
+        script {
+          sh('''
+            # If this is a job triggered by Gerrit because a new patchset has
+            # been published, abort all other older running builds that
+            # target the same change number
+            if [[ $GERRIT_EVENT_TYPE == 'patchset-created' ]]; then
+              python -u automation-git/scripts/jenkins/jenkins-job-cancel \
+                --older-than ${BUILD_NUMBER} \
+                --with-param GERRIT_CHANGE_NUMBER=${GERRIT_CHANGE_NUMBER} \
+                --wait 600 \
+                ${JOB_NAME} || :
+            else
+              if $voting; then
```

Comment:
So this else block will supercede existing jobs for the same GERRIT_CHANGE_NUMBER that are voting, or existing jobs for the same GERRIT_CHANGE_NUMBER that have the same gerrit_context... But what if there is a non-voting job that has a different gerrit_context? It would appear that that would be left running???

So this means that there can only be one running voting job for a GERRIT_CHANGE_NUMBER, but there could be many non-voting jobs for the same GERRIT_CHANGE_NUMBER so long as they use different gerrit_context values...

Reply:
Yes, that's precisely what I was trying to achieve. At most one voting job and many non-voting jobs, all with unique `gerrit_context` values. It's a bit difficult to explain. I hope the documentation does a better job of it.